lib/Common/DataStructures/DictionaryEntry.h

```diff
@@ -6,157 +6,143 @@
-        void Clear()
+        protected:
+            TValue value;    // data of entry
+        public:
+            int next;        // Index of next entry, -1 if last
+        };
+
+        template <class TKey, class TValue>
+        class KeyValueEntryDataLayout1
         {
-            this->value = nullptr;
-        }
-    };
-
-    template <>
-    class ValueEntry<bool>: public BaseValueEntry<bool>
-    {
-    public:
-        void Clear()
+        protected:
+            TValue value;    // data of entry
+            TKey key;        // key of entry
+        public:
+            int next;        // Index of next entry, -1 if last
+        };
+
+        template <class TKey, class TValue>
+        class KeyValueEntryDataLayout2
         {
-            this->value = false;
-        }
-    };
-
-    template <>
-    class ValueEntry<int>: public BaseValueEntry<int>
-    {
-    public:
-        void Clear()
+        protected:
+            TValue value;    // data of entry
+        public:
+            int next;        // Index of next entry, -1 if last
+        protected:
+            TKey key;        // key of entry
+        };
+
+        // Packing matters because we make so many dictionary entries.
+        // The int pointing to the next item in the list may be included
+        // either after the value or after the key, depending on which
+        // packs better.
+        template <class TKey, class TValue>
+        using KeyValueEntryData = ChooseSmaller<KeyValueEntryDataLayout1<TKey, TValue>, KeyValueEntryDataLayout2<TKey, TValue>>;
+
+        template <class TValue, class TData = ValueEntryData<TValue>>
+        class ValueEntry : public TData
         {
-            this->value = 0;
-        }
-    };
+        protected:
+            void Set(TValue const& value)
+            {
+                this->value = value;
+            }
+
+        public:
+            static bool SupportsCleanup()
+            {
+                return false;
+            }
+
+            static bool NeedsCleanup(ValueEntry<TValue, TData>&)
+            {
+                return false;
+            }
+
+            void Clear()
+            {
+                ClearValue<TValue>::Clear(&this->value);
+            }
+
+            TValue const& Value() const { return this->value; }
+            TValue& Value() { return this->value; }
+            void SetValue(TValue const& value) { this->value = value; }
+        };
+
+        // Used by BaseHashSet, the default is that the key is the same as the value
+        template <class TKey, class TValue>
+        class ImplicitKeyValueEntry : public ValueEntry<TValue>
+        {
+        public:
+            TKey Key() const { return ValueToKey<TKey, TValue>::ToKey(this->value); }
 
-    template <>
-    class ValueEntry<uint>: public BaseValueEntry<uint>
-    {
-    public:
-        void Clear()
+            void Set(TKey const& key, TValue const& value)
+            {
+                __super::Set(value);
```

Comment:
nit: I know we use __super in a few places in the repo already, but it'd be nice to have an implementation not using a vendor extension.

Reply:
Clang 3.6 added __super (no reason why multiple vendors can't be equally non-standard :) )

Reply:
The alternative would be something like `typedef ValueEntry<TValue, KeyValueEntryData<TKey, TValue>> super;` at the top of the class. It is true that it's available in our supported compilers though, and we do use it pretty extensively. It's a very nitpicky issue; the case where it's relevant is only if we end up wanting GCC support at some point, since as of GCC8.1 I think they still don't support __super.